File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/ruby/find_shortest_string_recursive.rb
def find_shortest_string_recursive(arr)
  # type your code in here
  return arr.first if arr.length == 1

  result = find_shortest_string_recursive(arr[1..-1])

  arr.first.length <= result.length ? result : result
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string_recursive(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string_recursive(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string_recursive(['flower', 'juniper', 'lily', 'dandelion'])

  puts("Expecting: 'cat'");
  puts("=>", find_shortest_string_recursive(['cat']))

  puts

  puts("Expecting: 'e'");
  puts("=>", find_shortest_string_recursive(['e','tt','ada','dasfs','ff']))

  puts

  puts("Expecting: 'cat'");
  puts("=>", find_shortest_string_recursive(['cat', 'lamb', 'elephant', 'ship']))

  puts

  puts("Expecting: 'ee'");
  puts("=>", find_shortest_string_recursive(['ee', 'eeee','eeeeee','eeeeeee']))

  puts

  puts("Expecting: ''");
  puts("=>", find_shortest_string_recursive(['','ds','adsf','ddda','as']))

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/javascript/find_shortest_string_recursive.js
function findShortestStringRecursive(arr) {
  // type your code here
  if (arr.length === 1) {   // if the array contains only one element, return the first element
    return arr[0]; 
  }

  const result = findShortestStringRecursive(arr.slice(1)); // recusrively use the function and pass in the array starting at indexthe first element 

  return arr[0].length <= result.length ? arr[0] : result; // ternary goes through the stack and compares each element and returns either the first element or the result of the function
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestStringRecursive(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestStringRecursive(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestStringRecursive(['flower', 'juniper', 'lily', 'dandelion']));

  console.log("");

  console.log("Expecting: 'cat'");
  console.log("=>", findShortestStringRecursive(['cat']));

  console.log("");

  console.log("Expecting: 'e'");
  console.log("=>", findShortestStringRecursive(['e','tt','ada','dasfs','ff']));

  console.log("");

  console.log("Expecting: 'cat'");
  console.log("=>", findShortestStringRecursive(['cat', 'lamb', 'elephant', 'ship']));

  console.log("");

  console.log("Expecting: 'ee'");
  console.log("=>", findShortestStringRecursive(['ee', 'eeee','eeeeee','eeeeeee']));

  console.log("");

  console.log("Expecting: ''");
  console.log("=>", findShortestStringRecursive(['','ds','adsf','ddda','as']));
}

module.exports = findShortestStringRecursive;

// Please add your pseudocode to this file
// And a written explanation of your solution
